### DISCUSSION TITLE
Defining global/static variables in CFS libraries.

### DISCUSSION THREAD

**Discussion by ryuhg** [nasa/cFS#364] [Q&A] [answered]
### Context

I have built a hardware-servicing library for a mission-specific piece of hardware, whose functions are intended to be called from a separate Application. This library uses a global "LibData" structure similar to the AppData in cFS Applications to store status, counters, etc.
I found that sometimes (but not always) the contents in this LibData are immediately changed after its initialization, seemingly to some garbage data. At this point I am not even sure which thread is responsible for it. My blind suspicion is that cFS library objects loaded by the core services are somewhat different from a typical .so objects and no static objects should be defined there.

The target platform uses armv5 Linux.

```
//Library Source
...
// global
MyLibData_t    LibData;
...
//Function Definitions
int32 MyLibInit(void) {
    MyLibData.MyBool = true;
    ...
}
...
```
```
//App Source Calling MyLib Functions
int32 MyAppInit(void) {
    ...
    status = MyLibInit();
    // At this point, for example, MyLibData.MyBool has a value 0.
    ...
```

### Question

Is it allowed to declare/define static variables in CFE_LIB objects? (Could these data be corrupted/dumped by other tasks?)

**Comment by ryuhg** [comment, marked as answer]
Sorry all. It was a dumb mistake. It turns out I missed there was another object named with an identical name elsewhere reserved for a test, which gets initialized just after the one I asked. Adding the static keyword eliminates the problem as it should.

**Comment by skliper** [comment]
Glad you figured it out!  The answer is "yes" it's allowed but you do need to avoid conflicts!
